File: desktop-tauri/src/ipc/accountSecurity.ts
import { invoke } from "@tauri-apps/api/core";

import type { AccountDevice, TwoFactorSetup, TwoFactorState } from "../types";

/** The second factor, and every device this account is signed in on. The
 * bearer token stays native: these name values, never routes. */

export function twoFactorStatus(): Promise<TwoFactorState> {
  return invoke("account_two_factor_status");
}

/** Starts a setup. The secret travels once, here, and never again. */
export function twoFactorStart(): Promise<TwoFactorSetup> {
  return invoke("account_two_factor_start");
}

/** Confirms the setup and returns the recovery codes: the only sight of them. */
export function twoFactorConfirm(code: string): Promise<string[]> {
  return invoke("account_two_factor_confirm", { code });
}

export function twoFactorReplaceRecoveryCodes(code: string): Promise<string[]> {
  return invoke("account_two_factor_recovery_codes", { code });
}

export function twoFactorDisable(code: string): Promise<void> {
  return invoke("account_two_factor_disable", { code });
}

/** Opens Apple's or Google's own security page, for an account whose second
 * step belongs to them. The renderer names a provider, never an address. */
export function accountProviderSecurity(provider: string): Promise<void> {
  return invoke("account_provider_security", { provider });
}

export function accountDevices(): Promise<AccountDevice[]> {
  return invoke("account_devices");
}

/** `signedOut` is true when the device signed out was this Mac. */
export function accountDeviceRevoke(device: string): Promise<{ signedOut: boolean }> {
  return invoke("account_device_revoke", { device });
}

export function accountDevicesRevokeAll(): Promise<{ signedOut: boolean }> {
  return invoke("account_devices_revoke_all");
}

export function accountDeleteWithPassword(password: string): Promise<void> {
  return invoke("account_delete_with_password", { password });
}

/** Resolves when the provider confirms, the attempt expires, or it is
 * cancelled from here — so it is awaited for as long as that takes. */
export function accountDeleteWithProvider(provider: string): Promise<void> {
  return invoke("account_delete_with_provider", { provider });
}

export function accountDeleteCancel(): Promise<void> {
  return invoke("account_delete_cancel");
}

File: desktop-tauri/src-tauri/src/commands/account_security.rs
use tauri::State;

use crate::account_delete;
use crate::account_devices::{self, AccountDevice, RevokeOutcome};
use crate::account_security::{self, TwoFactorSetup, TwoFactorState};
use crate::state::AppState;

#[tauri::command]
pub async fn account_two_factor_status(
    state: State<'_, AppState>,
) -> Result<TwoFactorState, String> {
    account_security::status(&state).await
}

#[tauri::command]
pub async fn account_two_factor_start(
    state: State<'_, AppState>,
) -> Result<TwoFactorSetup, String> {
    account_security::start(&state).await
}

/// Confirms a pending setup and answers with the recovery codes, which are
/// readable this once and never again.
#[tauri::command]
pub async fn account_two_factor_confirm(
    state: State<'_, AppState>,
    code: String,
) -> Result<Vec<String>, String> {
    account_security::confirm(&state, code).await
}

#[tauri::command]
pub async fn account_two_factor_recovery_codes(
    state: State<'_, AppState>,
    code: String,
) -> Result<Vec<String>, String> {
    account_security::replace_recovery_codes(&state, code).await
}

#[tauri::command]
pub async fn account_two_factor_disable(
    state: State<'_, AppState>,
    code: String,
) -> Result<(), String> {
    account_security::disable(&state, code).await
}

/// Opens the provider's own security page, for an account whose second step
/// belongs to Apple or Google rather than to Vibyra.
#[tauri::command]
pub fn account_provider_security(provider: String) -> Result<(), String> {
    account_security::open_provider_security(&provider)
}

#[tauri::command]
pub async fn account_devices(state: State<'_, AppState>) -> Result<Vec<AccountDevice>, String> {
    account_devices::list(&state).await
}

#[tauri::command]
pub async fn account_device_revoke(
    state: State<'_, AppState>,
    device: String,
) -> Result<RevokeOutcome, String> {
    account_devices::revoke_device(&state, device).await
}

#[tauri::command]
pub async fn account_devices_revoke_all(
    state: State<'_, AppState>,
) -> Result<RevokeOutcome, String> {
    account_devices::revoke_all(&state).await
}

/// Deletes an email account, proved by its password.
#[tauri::command]
pub async fn account_delete_with_password(
    state: State<'_, AppState>,
    password: String,
) -> Result<(), String> {
    account_delete::with_password(&state, password).await
}

/// Deletes an Apple or Google account by signing in with that provider once
/// more. Resolves when the provider confirms, the attempt expires, or the
/// person cancels; it is long-lived on purpose.
#[tauri::command]
pub async fn account_delete_with_provider(
    state: State<'_, AppState>,
    provider: String,
) -> Result<(), String> {
    account_delete::with_provider(&state, provider).await
}

#[tauri::command]
pub fn account_delete_cancel(state: State<'_, AppState>) {
    account_delete::cancel(&state);
}
